### ISSUE TITLE
What's the meaning of the model's output?How can I get the true distance from the model output?

### ISSUE THREAD

**Issue by duyuwen-duen** [DepthAnything/Depth-Anything-V2#93]
As it is mentioned in KITTI datasets(https://europe.naverlabs.com/research/computer-vision/proxy-virtual-worlds-vkitti-1/),the value of Depth values are distances to the camera plane obtained from the z-buffer that a pixel intensity of 1 in our single channel PNG16 depth images corresponds to a distance of 1cm to the camera plane.
So the depth value should be small when it's a close object.However,the depth value is big  when it's a close object in this method,which is totally different from the groundtruth.Could you tell me the reason and how can I get the true distance from the model output?

**Comment by Edric-star** [comment]
Hi, did you try the metric depth methods? It outputs the distances with units in meters. Please check it [here](https://github.com/DepthAnything/Depth-Anything-V2/blob/main/metric_depth/run.py).

**Comment by duyuwen-duen** [comment]
I find there is same problem with run.py.The close object is white,which means the value is big,but I think if the output of the model should like the groundtruth,whose close object should be black,because the depth value represents the distance?There is
![result](https://github.com/user-attachments/assets/c73d0a14-c265-40a8-9340-4e953e0425cc)
 the result I get.I think the result is reversed,but I don't not why.

**Comment by duyuwen-duen** [comment]
If I use the metric/run.py,there are the same problem.Besides that,how can I decide the value of max_depth?Looking forword to your reply!
![Uploading 1341848025.462832.png…]()

**Comment by Edric-star** [comment]
It seems like your used indoor pictures, please make sure that you used indoor hypersim model to predict

**Comment by duyuwen-duen** [comment]
If I use the outdoor pictures,it's also same.The value of the close object is bigger than far object,which is white.But if the depth value represents the distance,the value of the close object should be smaller than far object.
![demo02](https://github.com/user-attachments/assets/d498127b-de48-47ed-807a-e0cff8bce3cd)

**Comment by heyoeyo** [comment]
The (relative) models output depth that is like ~1 / true depth. So things that are far away output values close to zero (e.g. 1 / large number) and things that are very close output large values (e.g. 1 / small number), which is why the black/white values seem reversed.

The [paper](https://arxiv.org/abs/2406.09414) describes this output as "affine-invariant inverse depth" (see section **5.2** on page 6). The [v1 paper](https://arxiv.org/abs/2401.10891) describes it in a bit more detail in section **3.1** (page 3). I also have a description of it with some diagrams [here](https://github.com/heyoeyo/muggled_dpt/blob/main/.readme_assets/results_explainer.md).

**Comment by duyuwen-duen** [comment]
> Hi, did you try the metric depth methods? It outputs the distances with units in meters. Please check it [here](https://github.com/DepthAnything/Depth-Anything-V2/blob/main/metric_depth/run.py).

You are right!I use the old checkpoints to run the model,as soon as I change the checkpoints in metrics/README.md，I find it's right!

**Comment by duyuwen-duen** [comment]
> The (relative) models output depth that is like ~1 / true depth. So things that are far away output values close to zero (e.g. 1 / large number) and things that are very close output large values (e.g. 1 / small number), which is why the black/white values seem reversed.
> 
> The [paper](https://arxiv.org/abs/2406.09414) describes this output as "affine-invariant inverse depth" (see section **5.2** on page 6). The [v1 paper](https://arxiv.org/abs/2401.10891) describes it in a bit more detail in section **3.1** (page 3). I also have a description of it with some diagrams [here](https://github.com/heyoeyo/muggled_dpt/blob/main/.readme_assets/results_explainer.md).

Thank you for your reply!I haven't notice that,and I have been confused for a long time!Now I got it.Thank you very much!!

**Comment by bhack** [comment]
@duyuwen-duen the annoying thing is that also with the metric depth we need to specify the max distance x frame. Especially in a sequence where the camera is moving a lot over Z It could be a real pain.

**Comment by rhelck** [comment]
> > Hi, did you try the metric depth methods? It outputs the distances with units in meters. Please check it [here](https://github.com/DepthAnything/Depth-Anything-V2/blob/main/metric_depth/run.py).
> 
> You are right!I use the old checkpoints to run the model,as soon as I change the checkpoints in metrics/README.md，I find it's right!

In the run.py file, are the raw metric distances what are saved in the .npy file?

**Comment by duyuwen-duen** [comment]
> > > Hi, did you try the metric depth methods? It outputs the distances with units in meters. Please check it [here](https://github.com/DepthAnything/Depth-Anything-V2/blob/main/metric_depth/run.py).
> > 
> > 
> > You are right!I use the old checkpoints to run the model,as soon as I change the checkpoints in metrics/README.md，I find it's right!
> 
> In the run.py file, are the raw metric distances what are saved in the .npy file?

If you use the metric/run.py,and use the appropriate model in metric/README.md,you can gain a depth picture whose value represents the unit meters.

**Comment by rhelck** [comment]
I think I got it, its just the raw output unadjusted for 255 color spectrum? Setting the max depth helped quite a bit

**Comment by duyuwen-duen** [comment]
> I think I got it, its just the raw output unadjusted for 255 color spectrum? Setting the max depth helped quite a bit
You are right.If you use the model in metric/depth,the output of the model is represents the distance ,and if you want to show it in a picture,you can adjust it for 255 color spectrum.If you only want to use the value of the depth,you don't need to adjust for 255 color spectrum.

**Comment by DebbyX3** [comment]
Sorry for the reiteration, but I'm not sure if I'm getting it: 
the `.npy` files generated when running `metric_depth/run.py` contain the estimated depths from the camera in _meters_ for each pixel, correct? (ofc with the correct models/checkpoints suggested in the `metric_depth/README.md`)

So the image is just a normalized version of them to fit 0-255... Because gray value of pixel != value of array in the same point (also because the arrays contain floats)

**Comment by duyuwen-duen** [comment]
> Sorry for the reiteration, but I'm not sure if I'm getting it: the `.npy` files generated when running `metric_depth/run.py` contain the estimated depths from the camera in _meters_ for each pixel, correct? (ofc with the correct models/checkpoints suggested in the `metric_depth/README.md`)
> 
> So the image is just a normalized version of them to fit 0-255... Because gray value of pixel != value of array in the same point (also because the arrays contain floats)

![image](https://github.com/user-attachments/assets/f4321d29-3509-4180-aa04-054fe668a115)
![Uploading image.png…]()
Yes, you are correct. However, I have saved the depth data as a .png file, where the pixel values can represent distances. This means I have altered the run.py script, changing the significance of the .png file. For instance, if the depth value is 1.23, I add 1000 to it, resulting in a value of 1230, which I then store in a 16-bit format.But if you don't need it, you needn't to change it.

**Comment by szhang963** [comment]
Hi, I got a bad result when I used the metric/run.py and depth_anything_v2_metric_vkitti_vitl.pth to test the outdoor image.
![image](https://github.com/user-attachments/assets/b9ae8a7c-0b21-4ac5-85ae-58684da66048)
However, the result is right for using run.py.
![image](https://github.com/user-attachments/assets/7d3eca1c-44a2-48ac-b662-134b67dbffdd)
The original image is here.
![image](https://github.com/user-attachments/assets/64b9c37d-77d3-425a-bfcd-7e66b5064f95)

@duyuwen-duen, I have used your outdoor image, and I still got a bad result.
![image](https://github.com/user-attachments/assets/801caca8-8f18-4e21-bcf2-1d3663f8d45d)

Could someone provide some suggestions?

**Comment by Morland424** [comment]
> > I think I got it, its just the raw output unadjusted for 255 color spectrum? Setting the max depth helped quite a bit
> > You are right.If you use the model in metric/depth,the output of the model is represents the distance ,and if you want to show it in a picture,you can adjust it for 255 color spectrum.If you only want to use the value of the depth,you don't need to adjust for 255 color spectrum.

I have generated this depth map from the run.py, but I don't know how to extract exact depth value in meters from it, like it's just a coloured image. How so I get the depth in meters from this depth map?
![image](https://github.com/user-attachments/assets/0ee5ca8f-fd76-4c6d-9759-ea1f64001e47)

**Comment by heyoeyo** [comment]
@Morland424 You might have already found this by now, but just in case, the depth values will be available from the 'raw output' of the model after running the `depth_anything.infer_image(...)` part of the code. Assuming you run the metric depth run.py script, it would be [line 60](https://github.com/DepthAnything/Depth-Anything-V2/blob/31dc97708961675ce6b3a8d8ffa729170a4aa273/metric_depth/run.py#L60). And if you use the `--save-numpy` script argument, that value gets saved as:
`(filename of image)_raw_depth_meter.npy`
